### PR TITLE
🏫 Organizations and Roles

### PR DESCRIPTION
**Pull request by awtkns**
- Adds a superadmin (on the user column)
- Adds org table
- Adds table to associate users to orgs
- Users can make themselves admins on local now

### CODE REVIEW

File: next/prisma/schema.prisma [hunk 1]
```diff
@@ -51,11 +51,59 @@ model User {
...
+
+model UserRole {
+  id              String  @id @default(cuid())
+  user_id         String
+  organization_id String?
+  role_id         String
+
+  user         User          @relation(fields: [user_id], references: [id], onDelete: NoAction)
+  role         Role          @relation(fields: [role_id], references: [id], onDelete: NoAction)
+  Organization Organization? @relation(fields: [organization_id], references: [id], onDelete: NoAction)
```
Comment (asim-shrestha) [inline next/prisma/schema.prisma:98]: Should we cascade delete `onDelete` for user and organization?

File: next/prisma/schema.prisma [hunk 2]
```diff
@@ -51,11 +51,59 @@ model User {
...
 
   @@index([email])
   @@index([createDate])
 }
 
+model Organization {
+  id          String     @id @default(cuid())
+  name        String     @unique
+  description String?
+  UserRole    UserRole[]
```
Comment (asim-shrestha) [inline next/prisma/schema.prisma:64]: `user_roles`?

File: next/prisma/schema.prisma [hunk 3]
```diff
@@ -44,18 +44,55 @@ model User {
...
   image         String?
+  superAdmin    Boolean   @default(false) @map("super_admin")
 
   createDate DateTime @default(now())
 
   accounts Account[]
   sessions Session[]
   runs     Run[]
   Agent    Agent[]
+  UserRole UserRole[]
```
Comment (asim-shrestha) [inline next/prisma/schema.prisma:55]: a bit in-consistent with casing, and plurals

Should we call this userRoles?

File: next/prisma/schema.prisma [hunk 4]
```diff
@@ -44,18 +44,55 @@ model User {
...
 
   @@index([email])
   @@index([createDate])
 }
 
+model Organization {
+  id   String @id @default(cuid())
+  name String @unique
+
+  created_by  String
```
Comment (asim-shrestha) [inline next/prisma/schema.prisma:65]: this is a username? email?

Thoughts on making this point to a owner user?

**Comment by asim-shrestha** [comment]
sqlite build failure

File: next/src/hooks/useAuth.ts [hunk 5]
```diff
@@ -11,14 +12,26 @@ interface Auth {
...
 }
 
-export function useAuth({ protectedRoute } = { protectedRoute: false }): Auth {
+interface UseAuthOptions {
+  protectedRoute?: boolean;
+  isAllowed?: (user: Session) => boolean;
+}
+
+export function useAuth(
+  { protectedRoute, isAllowed }: UseAuthOptions = { protectedRoute: false, isAllowed: () => true }
```
Comment (asim-shrestha) [inline next/src/hooks/useAuth.ts:21]: We specify defaults but the fields are also nullable?

**Review by asim-shrestha — APPROVED** [review]

**Comment by awtkns** [comment]
> sqlite build failure

https://github.com/prisma/prisma/issues/2219 good old primsa gotta love it

**Pull request merged**